lib/streamlit/runtime/state/query_params.py

```diff
@@ -49,8 +50,15 @@ def __getitem__(self, key: str) -> str:
 
     # Type checking users should handle the string serialization themselves
     # We will accept any type for the list and serialize to str just in case
-    def __setitem__(self, key: str, value: Union[str, List[str]]) -> None:
-        if isinstance(value, list):
+    def __setitem__(
+        self, key: str, value: Union[str, Set[str], Tuple[str], List[str]]
+    ) -> None:
+        if isinstance(value, dict):
+            raise StreamlitAPIException(
+                f"You cannot set a query params key `{key}` to a dictionary `{value}`."
+            )
```

Comment:
I don't think we want to print the value of the dictionary here. The stacktrace + key name will already be more than enough for a developer to find the offending line of code, and the dict value is probably just noise